Source/WebCore/Modules/url-pattern/URLPatternComponent.cpp

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright (C) 2024 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "URLPatternComponent.h"
+
+#include "URLPatternCanonical.h"
+
+namespace WebCore {
+namespace URLPatternUtilities {
+
+URLPatternComponent::URLPatternComponent(String&& patternString, RegExp* regularExpression, Vector<String>&& groupNameList, bool hasRegexpGroups)
+    : m_patternString(WTFMove(patternString))
+    , m_regularExpression(regularExpression)
+    , m_groupNameList(WTFMove(groupNameList))
+    , m_hasRegexpGroups(hasRegexpGroups)
+{
+}
+
+const String& URLPatternComponent::getPatternString() const
+{
+    return WTFMove(m_patternString);
+}
+
+ExceptionOr<URLPatternComponent> URLPatternComponent::compile(VM& vm, StringView input, EncodingCallbackType type, const URLPatternUtilities::URLPatternStringOptions& options)
+{
+    auto maybePartList = URLPatternUtilities::URLPatternParser::parse(input, options, type);
+    if (maybePartList.hasException())
+        return maybePartList.releaseException();
+    Vector<Part> partList = maybePartList.releaseReturnValue();
+
+    auto regexAndNameList = generateRegexAndNameList(partList, options);
```

Comment:
I think this would be better as:
```
auto [regularExpressionString, nameList] = generateRegexAndNameList(partList, options);
```
It would avoid having a "moved-out" `regexAndNameList` in scope for the rest of the function, which is super error-prone.